include/inexor/vulkan-renderer/world/collision.hpp

```diff
@@ -0,0 +1,250 @@
+#pragma once
+
+#include <inexor/vulkan-renderer/world/cube.hpp>
+
+#include <glm/geometric.hpp>
+#include <glm/gtx/vector_angle.hpp>
+#include <glm/vec3.hpp>
+
+#include <array>
+#include <limits>
+#include <memory>
+#include <optional>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+namespace inexor::vulkan_renderer::world {
+
+/// @brief A wrapper for collisions between a ray and octree geometry.
+/// This class is used for octree collision, but it can be used for every cube-like data structure
+/// @tparam T A templatable type which offers a size() and center() method.
+template <typename T>
+class RayCubeCollision {
+    const T &m_cube;
+
+    using cube_face = std::tuple<std::string, glm::vec3>;
+    using cube_corner = std::tuple<std::string, glm::vec3>;
+    using cube_edge = std::tuple<std::string, glm::vec3>;
+
+    glm::vec3 m_intersection;
+    cube_face m_selected_face;
+    cube_corner m_nearest_corner;
+    cube_edge m_nearest_edge;
+
+public:
+    /// @brief Calculate point of intersection, selected face,
+    /// nearest corner on that face, and nearest edge on that face.
+    /// @param cube The cube to check for collision.
+    /// @param ray_pos The start point of the ray.
+    /// @param ray_dir The direction of the ray.
+    RayCubeCollision(const T &cube, const glm::vec3 ray_pos, const glm::vec3 ray_dir) : m_cube(cube) {
+
+        // This lambda adjusts the center points on a cube's face to the size of the octree,
+        // so collision works with cubes of any size. This does not yet account for rotations!
+        const auto adjust_coordinates = [=](const glm::vec3 pos) {
+            // TODO: Take rotation of the cube into account.
+            return m_cube.center() + pos * (m_cube.size() / 2);
+        };
```

Comment:
```suggestion
    RayCubeCollision(const T &cube, const glm::vec3 &ray_pos, const glm::vec3 &ray_dir) : m_cube(cube) {

        // This lambda adjusts the center points on a cube's face to the size of the octree,
        // so collision works with cubes of any size. This does not yet account for rotations!
        const auto adjust_coordinates = [&](const glm::vec3 &pos) {
            // TODO: Take rotation of the cube into account.
            return m_cube.center() + pos * (m_cube.size() / 2);
        };
```
No need to copy! Same for some other arguments below too.

Reply:
Done.